### PR TITLE
Peer to peer transactions

### PR DESCRIPTION
**Pull request by OrenZak**
#### Main purpose:
Allow the SDK client to create peer to peer transactions. 
#### Technical description:
- Added new API `Kin.payToUser(String offerJwt, @Nullable KinCallback<OrderConfirmation> callback)`. This API is similar to `purchase ` API.
- Added new exception code `ServiceException.USER_NOT_FOUND`
- Added a button to the sample app for testing `payToUser` API.
- Added Kotlin for sample app (where we are not using the SDK API eg: Dialogs)
#### Tests:
- Same tests as purchase API
#### Documentation:
- Added to source code and README.md
- Updated typos and links on README.md

### CODE REVIEW

File: app/src/main/java/com/ecosystem/kin/app/MainActivity.java [hunk 1]
```diff
@@ -79,30 +91,58 @@ public void onClick(View v) {
...
+						} else {
+							enableView(v, true);
+						}
+					}
+				});
+				dialog.show();
+			}
+		});
+
+		((TextView)findViewById(R.id.user_id_textview)).setText(getString(R.string.user_id, userID));
```
Comment (nitzantomer) [inline app/src/main/java/com/ecosystem/kin/app/MainActivity.java]: You're finding `R.id.user_id_textview` twice, why not:
```
final TextView textView = (TextView)findViewById(R.id.user_id_textview);
textView.setText(getString(R.string.user_id, userID));
textView.setOnClickListener(new OnClickListener() { ...
```

File: app/src/main/java/com/ecosystem/kin/app/MainActivity.java [hunk 2]
```diff
@@ -371,11 +443,22 @@ private void showToast(String msg) {
 		Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
 	}
 
+	private void showSnackbar(String msg, boolean isError) {
+		Snackbar snackbar = Snackbar.make(containerLayout, msg, isError ? Snackbar.LENGTH_LONG : Snackbar.LENGTH_SHORT);
+		if(isError) {
```
Comment (nitzantomer) [inline app/src/main/java/com/ecosystem/kin/app/MainActivity.java]: Missing space

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 3]
```diff
@@ -160,28 +160,32 @@ public Balance getBalance() {
...
 			@Override
 			public void onResult(final kin.core.Balance balanceObj) {
 				setBalance(balanceObj);
-				mainThread.execute(new Runnable() {
-					@Override
-					public void run() {
-						callback.onResponse(balance.getValue());
-					}
-				});
+				if(callback != null) {
```
Comment (nitzantomer) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: missing space

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 4]
```diff
@@ -160,28 +160,32 @@ public Balance getBalance() {
...
 
 			@Override
 			public void onError(final Exception e) {
-				mainThread.execute(new Runnable() {
-					@Override
-					public void run() {
-						callback.onFailure(ErrorUtil.getBlockchainException(e));
-					}
-				});
+				if(callback != null) {
```
Comment (nitzantomer) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: missing space

File: sdk/src/main/java/com/kin/ecosystem/Kin.java [hunk 5]
```diff
@@ -302,6 +302,20 @@ public static void requestPayment(String offerJwt, @Nullable KinCallback<OrderCo
 		OrderRepository.getInstance().requestPayment(offerJwt, callback);
 	}
 
+	/**
+	 * Allowing a user to pay to a different user for an offer defined within your app, using KIN.
+	 * This call might take time, due to transaction validation on the blockchain network.
+	 *
+	 * @param offerJwt Represents a 'Pay to user' offer in a JWT manner.
+	 * @param callback {@link OrderConfirmation} The result will be a failure or a succeed with a jwt confirmation.
```
Comment (nitzantomer) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java]: `or a success` not `succeed`

File: app/src/main/java/com/ecosystem/kin/app/MainActivity.java [hunk 6]
```diff
@@ -57,18 +66,21 @@
 	protected void onCreate(Bundle savedInstanceState) {
 		super.onCreate(savedInstanceState);
 		setContentView(R.layout.activity_main);
+		userID = SignInRepo.getUserId(getApplicationContext());
```
Comment (YohayBar) [inline app/src/main/java/com/ecosystem/kin/app/MainActivity.java]: userID isn not required within the onCreate logic. Maybe just provide getUserID method ?

File: README.md [hunk 7]
```diff
@@ -480,62 +480,143 @@ Once the user has completed the task associated with the Earn offer, you request
...
+    });
+} 
+catch (ClientException exception) {
+    exception.printStackTrace();
+}
+```
+    
+    
+<a name="CreateCustomPayToUserOffer"></a>
+### Creating a Custom Pay To User Offer ###
```
[the remark this answers is not in the document]
Comment (OrenZak) [reply README.md]: Can you suggest?

File: app/src/main/java/com/ecosystem/kin/app/MainActivity.java (hunk 6 above)
Comment (OrenZak) [reply app/src/main/java/com/ecosystem/kin/app/MainActivity.java]: not really important

File: core/src/main/java/com/kin/ecosystem/core/util/ErrorUtil.java [hunk 8]
```diff
@@ -40,16 +41,23 @@ public static KinEcosystemException fromApiException(ApiException apiException)
 				case 400:
 				case 401:
 				case 404:
+					if (apiException.getResponseBody() != null && apiException.getResponseBody().getCode() == 4046) {
```
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/util/ErrorUtil.java]: 4046 should be const as way to document what does 4046 code mean

File: sdk/src/main/java/com/kin/ecosystem/history/presenter/OrderHistoryPresenter.java [hunk 9]
```diff
@@ -53,6 +59,8 @@ public void onAttach(IOrderHistoryView view) {
 		eventLogger.send(OrderHistoryPageViewed.create());
 		getOrderHistoryList();
 		listenToCompletedOrders();
+		// Trigger to update balance, if someone paid to this user(p2p).
+		blockchainSource.getBalance(null);
```
Comment (YohayBar) [inline sdk/src/main/java/com/kin/ecosystem/history/presenter/OrderHistoryPresenter.java]: the balance presenter should handle this logic. When ever the balance presser is attached it should add observer to balance updated and remove it when it detached.

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (YohayBar) [reply README.md]: why not just payToUser ( App client side can decide if it needs to wait for blockchain conformation or move on the API is the same )

[the remark this answers is not in the document]
Comment (YohayBar) [reply README.md]: same as the api method name

**Pull request merged**